power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/basics.hpp

```diff
@@ -42,6 +42,63 @@ using RawOptions = PGM_Options;
 using RawDeserializer = PGM_Deserializer;
 using RawSerializer = PGM_Serializer;
 
+enum CalculationType {
```

Comment:
Here and elsewhere
```suggestion
enum class CalculationType {
```
Otherwise there's no point in providing them here

Reply:
But there is not direct conversion from C++ style `enum class` to C style `enum`, hence we could not directly use them (unless conversion functionality is provided in the member functions that expect them as arguments, which I don't think we should add to keep the wrapper as thin as possible). So I added it here so we could have it in the C++ API without having to call the C one, unless the user specifically wants it. However, now that I think about it, `basics.h` is already included in `basics.hpp`; hence, would it make sense to make aliases for the C `enums` just to align with the wrapper style of not using the `PGM_` prefix (to make a differentiation from the C API), or should I just leave it as is and make the user use such `enum`s as one would in C?

Reply:
I believe (not sure, please double-check) you could do something like

```cpp
enum class CalculationType {
    power_flow = PGM_power_flow
    // ...
};
```

but the implementation, you will have to do a static cast. The C++ interface should do this static cast itself.

The real annoying part is that bare `enum` types clog the namespace. you might just as well define `constexpr auto power_flow{PGM_power_flow};` without enums. A user will probably precisely want to use that enum class type-like functionality instead of raw values

if you really want to, you can do both, e.g.

```cpp
namespace power_grid_model_cpp {
namespace calculation_type {
using power_flow = PGM_power_flow; // acts like PGM_power_flow but in a namespace
} // namespace calculation_type

enum class CalculationType {
    power_flow = calculation_type::power_flow;
};
} // namespace power_grid_model_cpp
```

Reply:
Wouldn't doing something like
```
namespace calculation_type {
using power_flow = PGM_power_flow; // acts like PGM_power_flow but in a namespace
}
```

be sufficient? The `enum class` definition still has the problem of no direct conversion and that may be problematic and not so easy to handle in all cases. I start thinking that we might be overcomplicating things and maybe we should just leave the explicit `enum`s out from `basics.hpp` and allow the user to refer to them how they would do it at the C API level.

Reply:
@TonyXiang8787 do you have an opinion?

Reply:
I do not really see the added value of both proposal. The `PGM_*` enum in global scope is there anyway. We can just use them directly. I do not see the benefit to define another name inside namespace.

Reply:
Makes sense. I will remove this then.

Reply:
@mgovers @figueroa1395 

In my opinion the same can be applied to other places. We define something in a namespace in C++ in order to avoid naming conflict and confusion. But if the name is already defined in the global namespace as part of C-API (`PGM_*`), I highly hesitate to define another alias in the `pgm_c` namespace. This does not reduce the number of names in global namespace. This applies to types, enums, etc.

Reply:
@TonyXiang8787 You suggest that, for instance,
```
using Idx = PGM_Idx;
using ID = PGM_ID;

using PowerGridModel = PGM_PowerGridModel;
using MetaDataset = PGM_MetaDataset;
using MetaComponent = PGM_MetaComponent;
using MetaAttribute = PGM_MetaAttribute;
using RawConstDataset = PGM_ConstDataset;
using RawMutableDataset = PGM_MutableDataset;
using RawWritableDataset = PGM_WritableDataset;
using RawDatasetInfo = PGM_DatasetInfo;
using RawOptions = PGM_Options;
using RawDeserializer = PGM_Deserializer;
using RawSerializer = PGM_Serializer;
```

be removed altogether?

Reply:
Indeed it that's my suggestion. Consult also @mgovers

Reply:
For me, the thing is that when I'm using the C++ wrapper (e.g. because i'm writing tests), I would prefer to not have to write in every file `PGM_*` for every type, or `using PGM_*` in every test. I'm OK with including a separate header file with `using` declarations`, or one `using namespace power_grid_model_c::short_hands` (or similar) though, (but keep in mind that `using namespace` is usually discouraged nowadays for good reasons!!!).

One more thing to consider is that `enum` without `class` is not an actual type and therefore does not give additional compiler warnings when I'm doing stuff like providing the wrong input order. Errors like `set_calculation_type(options, PGM_newton_raphson)` are very easy to make and may go undectected for a long time. In tests in our repo, that would still be OK and probably be found relatively quickly, but if I were to be a user, I may not detect the mistake for a long time.

Reply:
Currently the `enum`s have been removed, but the basic "wrapper" types are still there. We can come back to this in the end to make the final call. I personally feel it is good as is, without cluttering everything much since we removed the `enum`s.

Reply:
I will leave you guys to decide.